src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newStates = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'addProperties':
+        const StateBfAdd = { ...(newStates[newStates.length - 1] || state) };
+
+        newStates.push({ ...StateBfAdd, ...action.extraData });
+        break;
+      case 'removeProperties':
+        const stateBfRemove = { ...(newStates[newStates.length - 1] || state) };
```

Comment:
[NAMING] - The variable name 'stateBfRemove' should be more descriptive and follow naming conventions. A better name would be 'stateBeforeRemove' or 'stateCopy'.